File: core/src/nl/t64/game/rpg/screens/inventory/equipslot/EquipSlot.java
package nl.t64.game.rpg.screens.inventory.equipslot;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.audio.AudioEvent;
import nl.t64.game.rpg.components.party.HeroItem;
import nl.t64.game.rpg.components.party.InventoryGroup;
import nl.t64.game.rpg.screens.inventory.itemslot.InventoryImage;
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot;
import nl.t64.game.rpg.screens.inventory.messagedialog.MessageDialog;
import nl.t64.game.rpg.screens.inventory.tooltip.ItemSlotTooltip;


class EquipSlot extends ItemSlot {

    private static final Color TRANSPARENT = new Color(1f, 1f, 1f, 0.3f);

    private final HeroItem heroItem;

    EquipSlot(int index, InventoryGroup filterGroup, ItemSlotTooltip tooltip, HeroItem heroItem) {
        super(index, filterGroup, tooltip);
        this.heroItem = heroItem;
        this.imagesBackground.addActorBefore(this.imagesBackground.getChildren().peek(),
                                             createShadowImage(this.filterGroup));
    }

    private static Image createShadowImage(InventoryGroup inventoryGroup) {
        String groupId = inventoryGroup.name().toLowerCase();
        TextureRegion textureRegion = Utils.getResourceManager().getAtlasTexture(groupId);
        Image shadow = new Image(textureRegion);
        shadow.setColor(TRANSPARENT);
        shadow.setName("shadow");
        return shadow;
    }

    @Override
    public void addToStack(Actor actor) {
        super.add(actor);
        if (actor instanceof InventoryImage inventoryImage) {
            heroItem.forceSetInventoryItemFor(filterGroup, inventoryImage.inventoryItem);
            refreshSlot();
        }
    }

    @Override
    public boolean hasItem() {
        return heroItem.getInventoryItem(filterGroup).isPresent();
    }

    @Override
    public boolean isOnHero() {
        return true;
    }

    @Override
    public boolean doesAcceptItem(InventoryImage draggedItem) {
        if (filterGroup.equals(draggedItem.inventoryGroup)) {
            return doesHeroAcceptItem(draggedItem);
        }
        return false;
    }

    @Override
    public void clearStack() {
        if (hasItem()) {
            super.getChildren().pop();
            heroItem.clearInventoryItemFor(filterGroup);
            refreshSlot();
        }
    }

    @Override
    public void putInSlot(InventoryImage draggedItem) {
        addToStack(draggedItem);
    }

    @Override
    public int getAmount() {
        return hasItem() ? 1 : 0;
    }

    @Override
    public void incrementAmountBy(int amount) {
        throw new IllegalCallerException("EquipSlot amount cannot be incremented.");
    }

    @Override
    public void decrementAmountBy(int amount) {
        heroItem.clearInventoryItemFor(filterGroup);
        refreshSlot();
    }

    private boolean doesHeroAcceptItem(InventoryImage draggedItem) {
        return heroItem.createMessageIfNotAbleToEquip(draggedItem.inventoryItem)
                       .map(this::showDialogHeroDoesNotAccept)
                       .orElse(true);
    }

    private boolean showDialogHeroDoesNotAccept(String message) {
        new MessageDialog(message).show(getStage(), AudioEvent.SE_MENU_ERROR);
        return false;
    }

    private void refreshSlot() {
        setVisibilityOfShadow();
        setItemColor();
    }

    private void setVisibilityOfShadow() {
        final boolean shouldShadowBeVisible = !hasItem();
        setShadowVisible(shouldShadowBeVisible);
    }

    private void setShadowVisible(boolean visible) {
        imagesBackground.getChild(1).setVisible(visible);  // 1 is index of actor named shadow
    }

}

File: core/src/nl/t64/game/rpg/screens/inventory/equipslot/EquipSlot.kt
package nl.t64.game.rpg.screens.inventory.equipslot

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.ui.Image
import nl.t64.game.rpg.Utils.resourceManager
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.components.party.HeroItem
import nl.t64.game.rpg.components.party.inventory.InventoryGroup
import nl.t64.game.rpg.screens.inventory.itemslot.InventoryImage
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot
import nl.t64.game.rpg.screens.inventory.messagedialog.MessageDialog
import nl.t64.game.rpg.screens.inventory.tooltip.ItemSlotTooltip


private val TRANSPARENT = Color(1f, 1f, 1f, 0.3f)

private fun createShadowImage(inventoryGroup: InventoryGroup): Image {
    val groupId = inventoryGroup.name.lowercase()
    val textureRegion = resourceManager.getAtlasTexture(groupId)
    return Image(textureRegion).apply {
        color = TRANSPARENT
        name = "shadow"
    }
}

internal class EquipSlot(
    index: Int,
    filterGroup: InventoryGroup,
    tooltip: ItemSlotTooltip,
    private val heroItem: HeroItem
) : ItemSlot(index, filterGroup, tooltip) {

    init {
        imagesBackground.addActorBefore(imagesBackground.children.peek(), createShadowImage(filterGroup))
    }

    override fun addToStack(actor: Actor) {
        super.add(actor);
        if (actor is InventoryImage) {
            heroItem.forceSetInventoryItemFor(filterGroup, actor.inventoryItem);
            refreshSlot();
        }
    }

    override fun hasItem(): Boolean {
        return heroItem.getInventoryItem(filterGroup) != null
    }

    override fun isOnHero(): Boolean {
        return true
    }

    override fun doesAcceptItem(draggedItem: InventoryImage): Boolean {
        return if (filterGroup == draggedItem.inventoryGroup) doesHeroAcceptItem(draggedItem) else false
    }

    override fun clearStack() {
        if (hasItem()) {
            super.getChildren().pop()
            heroItem.clearInventoryItemFor(filterGroup)
            refreshSlot()
        }
    }

    override fun putInSlot(draggedItem: InventoryImage) {
        addToStack(draggedItem)
    }

    override fun getAmount(): Int {
        return if (hasItem()) 1 else 0
    }

    override fun incrementAmountBy(amount: Int) {
        throw IllegalCallerException("EquipSlot amount cannot be incremented.")
    }

    override fun decrementAmountBy(amount: Int) {
        throw IllegalCallerException("EquipSlot amount cannot be decremented.")
    }

    private fun doesHeroAcceptItem(draggedItem: InventoryImage): Boolean {
        val message = heroItem.createMessageIfNotAbleToEquip(draggedItem.inventoryItem)
        return message == null || showDialogHeroDoesNotAccept(message)
    }

    private fun showDialogHeroDoesNotAccept(message: String): Boolean {
        MessageDialog(message).show(stage, AudioEvent.SE_MENU_ERROR)
        return false
    }

    private fun refreshSlot() {
        setVisibilityOfShadow()
        setItemColor()
    }

    private fun setVisibilityOfShadow() {
        val shouldShadowBeVisible = !hasItem()
        setShadowVisible(shouldShadowBeVisible)
    }

    private fun setShadowVisible(visible: Boolean) {
        imagesBackground.getChild(1).isVisible = visible // 1 is index of actor named shadow
    }

}
